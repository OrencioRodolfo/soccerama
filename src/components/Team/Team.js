import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import Button from 'material-ui/Button'
import Dialog, {
  DialogActions,
  DialogContent,
  DialogTitle,
  withMobileDialog,
} from 'material-ui/Dialog'
import { CircularProgress } from 'material-ui/Progress'
import styled from 'styled-components'
import actions from '../../store/actions'
import Card from '../Card/Card'

const LogoContainer = styled.div`
  display: flex;
  justify-content: center;
  margin-bottom: 20px;
`
const Squad = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  align-items: stretch;
`
const Player = styled.div`
  flex-grow: 1;
  display: flex;
  flex-direction: column;
`
const HeadContent = styled.div`
  display: flex;
  justify-content: space-between;
`
const Ul = styled.ul`
  padding: 0;
  margin: 0;
`
const Li = styled.li`
  padding: 0;
  margin: 0;
  list-style: none;
  display: flex;
  justify-content: space-between;
  border-bottom: 1px solid #d6d6d6;
  margin: 10px 0;
`
const Title = styled.h2`
  color: rgba(255, 255, 255, 0.87);
  margin: 0;
`
const styles = {
  dialogHead: {
    background: '#009688',
    color: 'rgba(255, 255, 255, 0.87)',
  },
}

class TeamDialog extends Component {
  static propTypes = {
    getPlayer: PropTypes.func.isRequired,
    fullScreen: PropTypes.bool.isRequired,
    open: PropTypes.bool,
    onClose: PropTypes.func,
    team: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      logo: PropTypes.string,
    }).isRequired,
    squad: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      logo: PropTypes.string,
    })).isRequired,
    players: PropTypes.shape({
      id: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        height: PropTypes.string,
        weight: PropTypes.string,
        photo: PropTypes.string,
      }),
    }).isRequired,
  }

  static defaultProps = {
    open: false,
    onClose: undefined,
  }

  state = {
    open: false,
    squadLoaded: false,
  }

  componentWillReceiveProps(props) {
    this.setState({ open: props.open })
  }

  getSquad = () => {
    if (this.props.squad) {
      const requests = this.props.squad.reduce((promises, player) => {
        if (!this.props.players[player.player_id]) {
          return [...promises, this.props.getPlayer(player.player_id)]
        }
        return []
      }, [])

      return Promise.all(requests).then(() => {
        this.setState({
          squadLoaded: true,
        })
      })
    }

    return null
  }

  handleClose = () => {
    this.setState({
      open: false,
      squadLoaded: false,
    })
    this.props.onClose()
  }

  renderSquad = () => {
    if (this.state.squadLoaded) {
      if (this.props.squad.length) {
        return this.props.squad.map((squadMember) => {
          const player = this.props.players[squadMember.player_id]

          return (
            player &&
            <Player key={player.id}>
              <Card>
                <Card.Header>
                  <HeadContent>
                    <img src={player.photo} alt="player" />
                    <b>{player.name}</b>
                  </HeadContent>
                </Card.Header>
                <Card.Content>
                  <Ul>
                    {player.position && <Li><b>Position:</b> {player.position}</Li>}
                    {player.weight && <Li><b>Weight:</b> {player.weight}</Li>}
                    {player.height && <Li><b>Height:</b> {player.height}</Li>}
                    {player.nationality && <Li><b>Nationality:</b> {player.nationality}</Li>}
                  </Ul>
                </Card.Content>
              </Card>
            </Player>
          )
        })
      }

      return (
        <h3>Squad not available</h3>
      )
    }

    return (
      <CircularProgress />
    )
  }

  render() {
    const { fullScreen } = this.props

    return (
      <div >
        <Dialog
          fullScreen={fullScreen}
          open={this.state.open}
          onClose={this.handleClose}
          onEnter={this.getSquad}
          aria-labelledby="Team squad"
          maxWidth="md"
          fullWidth
        >
          <DialogTitle style={styles.dialogHead} disableTypography>
            <Title>{this.props.team.name}</Title>
          </DialogTitle>
          <DialogContent style={{ padding: '15px' }}>
            <LogoContainer>
              <img src={this.props.team.logo} alt="team logo" />
            </LogoContainer>
            <Squad>{this.renderSquad()}</Squad>
          </DialogContent>
          <DialogActions>
            <Button onClick={this.handleClose} color="primary" autoFocus>
              Dismiss
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    )
  }
}

const mapStateToProps = ({ players }) => ({
  players,
})

export default connect(mapStateToProps, {
  getPlayer: actions.getPlayer,
})(withMobileDialog()(TeamDialog))

export {
  TeamDialog as DTeam,
  mapStateToProps,
}
