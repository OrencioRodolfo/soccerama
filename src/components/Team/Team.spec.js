import React from 'react'
import { shallow } from 'enzyme'
import { DTeam, mapStateToProps } from './Team'

describe('Home component', () => {
  let wrapper
  let instance
  let getPlayer
  let onClose
  let team
  let squad
  let players

  beforeEach(() => {
    getPlayer = () => Promise.resolve()
    onClose = jest.fn()
    team = {}
    squad = [{
      player_id: 1,
      name: 'sample',
      logo: 'logo',
    },
    {
      player_id: 2,
      name: 'sample',
      logo: 'logo',
    }]
    players = {
      1: {
        id: 1,
        name: 'Sample',
        height: 'Sample',
        weight: 'Sample',
        photo: 'Sample',
      },
    }

    wrapper = shallow(<DTeam
      getPlayer={getPlayer}
      fullScreen={false}
      open
      onClose={onClose}
      team={team}
      squad={squad}
      players={players}
    />)
    instance = wrapper.instance()
  })

  it('must render properly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('componentWillReceiveProps', () => {
    it('must set the dialog as "open"', () => {
      instance.componentWillReceiveProps({ open: true })
      expect(instance.state.open).toEqual(true)
    })
  })

  describe('fetch squad', () => {
    it('must fetch the entire squad', async () => {
      instance.state = {
        squadLoaded: false,
      }

      await instance.getSquad().then(() => {
        expect(instance.state.squadLoaded).toEqual(true)
      })
    })

    it('must render the entire squad', () => {
      instance.state = {
        squadLoaded: true,
      }

      expect(instance.renderSquad()).toMatchSnapshot()
    })
  })

  describe('render squad', () => {
    it('must render a circular progress while the squad isn\'t entirely loaded', () => {
      instance.state = {
        squadLoaded: false,
      }

      expect(instance.renderSquad()).toMatchSnapshot()
    })

    it('must render the entire squad', () => {
      instance.state = {
        squadLoaded: true,
      }

      expect(instance.renderSquad()).toMatchSnapshot()
    })
  })

  describe('dialog closing', () => {
    it('must update the state setting the dialog as closed', () => {
      instance.handleClose()
      expect(instance.state.open).toEqual(false)
      expect(instance.state.squadLoaded).toEqual(false)
      expect(onClose).toBeCalled()
    })
  })

  describe('mapStateToProps function', () => {
    it('must return the data related to the players', () => {
      const state = {
        players,
      }
      const result = mapStateToProps(state)
      expect(result).toHaveProperty('players')
      expect(result).toMatchSnapshot()
    })
  })
})
