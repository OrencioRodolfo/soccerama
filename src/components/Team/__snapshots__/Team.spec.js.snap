// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component fetch squad must render the entire squad 1`] = `
Array [
  <styled.div>
    <img
      alt="player"
      src="Sample"
    />
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Sample
    </div>
  </styled.div>,
  undefined,
]
`;

exports[`Home component mapStateToProps function must return the data related to the players 1`] = `
Object {
  "players": Object {
    "1": Object {
      "height": "Sample",
      "id": 1,
      "name": "Sample",
      "photo": "Sample",
      "weight": "Sample",
    },
  },
}
`;

exports[`Home component must render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TeamDialog
    fullScreen={false}
    getPlayer={[Function]}
    onClose={[MockFunction]}
    open={true}
    players={
      Object {
        "1": Object {
          "height": "Sample",
          "id": 1,
          "name": "Sample",
          "photo": "Sample",
          "weight": "Sample",
        },
      }
    }
    squad={
      Array [
        Object {
          "logo": "logo",
          "name": "sample",
          "player_id": 1,
        },
        Object {
          "logo": "logo",
          "name": "sample",
          "player_id": 2,
        },
      ]
    }
    team={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <WithStyles(Dialog)
        aria-labelledby="responsive-dialog-title"
        fullScreen={false}
        onClose={[Function]}
        onEnter={[Function]}
        open={false}
      >
        <WithStyles(DialogTitle) />
        <WithStyles(DialogContent)
          style={
            Object {
              "minWidth": "400px",
            }
          }
        >
          <styled.div>
            <img
              alt="team logo"
              src={undefined}
            />
          </styled.div>
          <styled.div>
            <WithStyles(CircularProgress) />
          </styled.div>
        </WithStyles(DialogContent)>
        <WithStyles(DialogActions)>
          <WithStyles(Button)
            autoFocus={true}
            color="primary"
            onClick={[Function]}
          >
            Dismiss
          </WithStyles(Button)>
        </WithStyles(DialogActions)>
      </WithStyles(Dialog)>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "aria-labelledby": "responsive-dialog-title",
        "children": Array [
          <WithStyles(DialogTitle) />,
          <WithStyles(DialogContent)
            style={
              Object {
                "minWidth": "400px",
              }
            }
          >
            <styled.div>
              <img
                alt="team logo"
                src={undefined}
              />
            </styled.div>
            <styled.div>
              <WithStyles(CircularProgress) />
            </styled.div>
          </WithStyles(DialogContent)>,
          <WithStyles(DialogActions)>
            <WithStyles(Button)
              autoFocus={true}
              color="primary"
              onClick={[Function]}
            >
              Dismiss
            </WithStyles(Button)>
          </WithStyles(DialogActions)>,
        ],
        "fullScreen": false,
        "onClose": [Function],
        "onEnter": [Function],
        "open": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.div>
                <img
                  alt="team logo"
                  src={undefined}
                />
              </styled.div>,
              <styled.div>
                <WithStyles(CircularProgress) />
              </styled.div>,
            ],
            "style": Object {
              "minWidth": "400px",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <img
                  alt="team logo"
                  src={undefined}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "team logo",
                  "src": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <WithStyles(CircularProgress) />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(Button)
              autoFocus={true}
              color="primary"
              onClick={[Function]}
            >
              Dismiss
            </WithStyles(Button)>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoFocus": true,
              "children": "Dismiss",
              "color": "primary",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Dismiss",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <WithStyles(Dialog)
          aria-labelledby="responsive-dialog-title"
          fullScreen={false}
          onClose={[Function]}
          onEnter={[Function]}
          open={false}
        >
          <WithStyles(DialogTitle) />
          <WithStyles(DialogContent)
            style={
              Object {
                "minWidth": "400px",
              }
            }
          >
            <styled.div>
              <img
                alt="team logo"
                src={undefined}
              />
            </styled.div>
            <styled.div>
              <WithStyles(CircularProgress) />
            </styled.div>
          </WithStyles(DialogContent)>
          <WithStyles(DialogActions)>
            <WithStyles(Button)
              autoFocus={true}
              color="primary"
              onClick={[Function]}
            >
              Dismiss
            </WithStyles(Button)>
          </WithStyles(DialogActions)>
        </WithStyles(Dialog)>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "aria-labelledby": "responsive-dialog-title",
          "children": Array [
            <WithStyles(DialogTitle) />,
            <WithStyles(DialogContent)
              style={
                Object {
                  "minWidth": "400px",
                }
              }
            >
              <styled.div>
                <img
                  alt="team logo"
                  src={undefined}
                />
              </styled.div>
              <styled.div>
                <WithStyles(CircularProgress) />
              </styled.div>
            </WithStyles(DialogContent)>,
            <WithStyles(DialogActions)>
              <WithStyles(Button)
                autoFocus={true}
                color="primary"
                onClick={[Function]}
              >
                Dismiss
              </WithStyles(Button)>
            </WithStyles(DialogActions)>,
          ],
          "fullScreen": false,
          "onClose": [Function],
          "onEnter": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <styled.div>
                  <img
                    alt="team logo"
                    src={undefined}
                  />
                </styled.div>,
                <styled.div>
                  <WithStyles(CircularProgress) />
                </styled.div>,
              ],
              "style": Object {
                "minWidth": "400px",
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <img
                    alt="team logo"
                    src={undefined}
                  />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "team logo",
                    "src": undefined,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <WithStyles(CircularProgress) />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <WithStyles(Button)
                autoFocus={true}
                color="primary"
                onClick={[Function]}
              >
                Dismiss
              </WithStyles(Button)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "autoFocus": true,
                "children": "Dismiss",
                "color": "primary",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Dismiss",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Home component render squad must render a circular progress while the squad isn't entirely loaded 1`] = `<WithStyles(CircularProgress) />`;

exports[`Home component render squad must render the entire squad 1`] = `
Array [
  <styled.div>
    <img
      alt="player"
      src="Sample"
    />
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Sample
    </div>
  </styled.div>,
  undefined,
]
`;
